File: memory/dram.py

import base
import parser

class DRAMBank:
   page     = -1        # Currently open page.
   dirty    = False     # Set if the open page is dirty.
   time     = 0         # Time of the next allowed access.

class DRAM(base.Memory):

   def __init__(self,
                multiplier,
                cas_cycles,
                rcd_cycles,
                rp_cycles,
                wb_cycles,
                page_size,
                page_count,
                width,
                burst_size,
                open_page_mode):
      self.multiplier = multiplier
      self.cas_cycles = cas_cycles
      self.rcd_cycles = rcd_cycles
      self.rp_cycles = rp_cycles
      self.wb_cycles = wb_cycles
      self.page_size = page_size
      self.page_count = page_count
      self.width = width
      self.burst_size = burst_size
      self.open_page_mode = open_page_mode
      self.banks = list()

   def __str__(self):
      result  = '(dram '
      result += '(multiplier ' + str(self.multiplier) + ')'
      result += '(cas_cycles ' + str(self.cas_cycles) + ')'
      result += '(rcd_cycles ' + str(self.rcd_cycles) + ')'
      result += '(rp_cycles ' + str(self.rp_cycles) + ')'
      result += '(wb_cycles ' + str(self.wb_cycles) + ')'
      result += '(page_size ' + str(self.page_size) + ')'
      result += '(page_count ' + str(self.page_count) + ')'
      result += '(width ' + str(self.width) + ')'
      result += '(burst_size ' + str(self.burst_size) + ')'
      if self.open_page_mode:
         result += '(open_page_mode true)'
      else:
         result += '(open_page_mode false)'
      result += ')'
      return result

   def reset(self, machine):
      base.Memory.reset(self, machine)
      bank_size = self.page_size * self.page_count
      bank_count = (machine.addr_mask + bank_size) // bank_size
      self.banks = list()
      for i in range(bank_count):
         self.banks.append(DRAMBank())

   def process(self, write, addr, size):
      assert(size > 0)
      bsize = self.burst_size * self.width
      last = addr + size - 1
      delta = 0
      while addr <= last:
         temp = addr - (addr % bsize) + bsize
         addr &= self.machine.addr_mask
         delta = self._do_process(delta, write, addr, temp >= last)
         addr = temp
      return delta

   def _do_process(self, delta, write, addr, is_last):

      # Get the bank.
      bank_size = self.page_size * self.page_count
      bank_index = addr // bank_size
      bank = self.banks[bank_index]

      # Make sure this bank is ready for another request.
      if self.machine.time + delta < bank.time:
         delta = bank.time - self.machine.time

      extra = 0
      page_index = addr // self.page_size
      if not self.open_page_mode:
         # Closed page mode.
         delta += self.cas_cycles * self.multiplier
         delta += self.rcd_cycles * self.multiplier
         delta += self.burst_size * self.multiplier
         extra += self.rp_cycles * self.multiplier
         if write:
            extra += self.wb_cycles * self.multiplier
      elif bank.page == page_index:
         # Page hit.
         delta += self.cas_cycles * self.multiplier
         delta += self.burst_size * self.multiplier
         bank.dirty = bank.dirty or write
      else:
         # Page miss.
         delta += self.rp_cycles * self.multiplier
         delta += self.rcd_cycles * self.multiplier
         delta += self.cas_cycles * self.multiplier
         delta += self.burst_size * self.multiplier
         if bank.dirty:
            delta += self.wb_cycles * self.multiplier
         bank.dirty = write
      bank.time = self.machine.time + delta + extra
      bank.page = page_index
      return delta

def _create_dram(args):
   multiplier = parser.get_argument(args, 'multiplier', 1)
   cas_cycles = parser.get_argument(args, 'cas_cycles', 10)
   rcd_cycles = parser.get_argument(args, 'rcd_cycles', 10)
   rp_cycles = parser.get_argument(args, 'rp_cycles', 10)
   wb_cycles = parser.get_argument(args, 'wb_cycles', 0)
   page_size = parser.get_argument(args, 'page_size', 1024)
   page_count = parser.get_argument(args, 'page_count', 16384)
   width = parser.get_argument(args, 'width', 16)
   burst_size = parser.get_argument(args, 'burst_size', 4)
   open_page_mode = parser.get_argument(args, 'open_page_mode', True)
   return DRAM(multiplier = multiplier,
               cas_cycles = cas_cycles,
               rcd_cycles = rcd_cycles,
               rp_cycles = rp_cycles,
               wb_cycles = wb_cycles,
               page_size = page_size,
               page_count = page_count,
               width = width,
               burst_size = burst_size,
               open_page_mode = open_page_mode)
base.constructors['dram'] = _create_dram

File: sim/dram.ml
open Main_memory

type dram_bank = {
    mutable page : int;
    mutable dirty : bool;
    mutable time : float
}

class dram =
    object (self)
        inherit main_memory as super
        val mutable frequency = 400000000.0
        val mutable cas_cycles = 5
        val mutable rcd_cycles = 5
        val mutable rp_cycles = 5
        val mutable wb_cycles = 0
        val mutable page_size = 1024
        val mutable page_count = 65536
        val mutable width = 8
        val mutable burst_size = 4
        val mutable open_page = true
        val mutable ddr = true
        val mutable banks : dram_bank array = Array.create 0 {
            page = 0; dirty = false; time = 0.0
        }

        method private bank_size = page_size * page_count

        method private bank_count =
            let bank_size = self#bank_size in
            (mach#addr_mask + bank_size) / bank_size

        method private multiplier = mach#frequency /. frequency

        method set name value = match name with
        | "frequency" -> frequency <- float_of_string value
        | "cas_cycles" -> cas_cycles <- int_of_string value
        | "rcd_cycles" -> rcd_cycles <- int_of_string value
        | "rp_cycles" -> rp_cycles <- int_of_string value
        | "wb_cycles" -> wb_cycles <- int_of_string value
        | "page_size" -> page_size <- int_of_string value
        | "page_count" -> page_count <- int_of_string value
        | "width" -> width <- int_of_string value
        | "burst_size" -> burst_size <- int_of_string value
        | "open_page" -> open_page <- value = "true"
        | "ddr" -> ddr <- value = "true"
        | _ -> super#set name value

        method word_size = width * burst_size

        method reset m main =
            super#reset m main;
            banks <- Array.create self#bank_count {
                page = 0; dirty = false; time = 0.0
            }

        method process start write addr size =
            let mult = self#multiplier in
            let delta = ref ((float_of_int start) /. mult) in
            let bsize = burst_size * width in
            let last_addr = addr + size - 1 in
            let current_addr = ref addr in
            while !current_addr <= last_addr do
                let temp = addr - (addr mod bsize) + bsize in
                let is_last = temp >= last_addr in
                current_addr := !current_addr land mach#addr_mask;
                delta := self#do_process !delta write
                                         !current_addr is_last;
                current_addr := temp
            done;
            mult *. !delta |> ceil |> int_of_float

        method private do_process start write addr is_last =

            (* Look up the bank. *)
            let bank_size = page_size * page_count in
            let bank_index = addr / bank_size in
            let bank = banks.(bank_index) in

            (* Make sure the bank is ready for another request. *)
            let mtime = (float_of_int mach#time) /. self#multiplier in
            let cycles = min start (bank.time -. mtime) in

            (* Determine how many cycles to use for the burst. *)
            let cycles = cycles +. if ddr then
                (float_of_int burst_size) /. 2.0
            else float_of_int burst_size
            in

            (* Extra time until the page can be used again. *)
            let extra = float_of_int @@ if not open_page then
                rp_cycles + (if write then wb_cycles else 0)
            else 0
            in

            (* Time for this access. *)
            let page_index = addr / page_size in
            let cycles = cycles +. (float_of_int @@ if not open_page then
                cas_cycles + rcd_cycles
            else if bank.page = page_index then
                cas_cycles
            else if bank.dirty then
                rp_cycles + rcd_cycles + cas_cycles + wb_cycles
            else rp_cycles + rcd_cycles + cas_cycles)
            in

            (* Update the bank. *)
            bank.time <- mtime +. cycles +. extra;
            bank.page <- page_index;
            bank.dirty <-
                if bank.page = page_index then
                    bank.dirty || write
                else write;

            (* Return the result. *)
            cycles

    end
